### DISCUSSION TITLE
Proposal: Allow yield return in anonymous functions

### DISCUSSION THREAD

**Discussion by hez2010** [dotnet/csharplang#2583] [Language Ideas]
Allow yield return in an anonymous function:
```csharp
var func = new Func<int, IEnumerable<int>>((n) => 
{
    for (var i = 0; i < n; i++) yield return i;
});

foreach (var i in func(1000)) .......
```

**Comment by ufcpp** [comment]
See https://github.com/dotnet/roslyn/issues/24#issuecomment-70448545, https://github.com/dotnet/roslyn/issues/24#issuecomment-239875217

> This scenario is better served with local functions, which can be iterators.

**Comment by CyrusNajmabadi** [comment]
Specifically, you can just take the above and write it as:

```c#
IEnumerable<int> func(int n) 
{
    for (var i = 0; i < n; i++) yield return i;
};

foreach (var i in func(1000)) .......
```

**Comment by mdealer** [comment]
It all depends on the specific case! From a bystanders perspective not supporting this in a certain corner of the language makes it inconsistent. 

I've got a system that I could fix by adding 10 lines of code with this, but now I have to resort changing APIs, introducing new classes etc. That's very 1980s, especially considering C# been getting more and more functional over time.

Regarding 'hardtoreads' and 'hardtounderstands' from the linked issue, it's way too late to be afraid of this for C# (or most languages that support latest buzz like anonymous, interfaces with implementation, async etc). If you ask me, the big amounts of classes that I need to add make it harder to read.

**Comment by CyrusNajmabadi** [comment]
> I've got a system that I could fix by adding 10 lines of code with this, but now I have to resort changing APIs, introducing new classes etc.

Why?

**Comment by mdealer** [comment]
@CyrusNajmabadi I just re-read your above post. Sure, I can define a local function. That's a brain-fuck right there. Why can't this be an anonymous function in the first place? It's inconsistent. I just want to assign the function to a Func<> variable, but can't. It's non-intuitive.

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi What do you mean? If I can't create an anonymous function that creates objects while iterating anywhere to pass it as argument, then I have to look for a class to put it in. 

No, you don' tneed a class.  There are lots of options.  You can have a sibling method/property.  Or you can have a local function.

>  That can have consequences.

I'm not sure what hte consequences would be of this.  A local function is just as capable here, and doesn't require any sort of classes to be created.

**Comment by CyrusNajmabadi** [comment]
> Why can't this be an anonymous function in the first place?

Because it hasn't been important enough to support that niche case when there are suitable alternatives already available in the language.  Resources are limited, and they're being allocated ot language features felt to be more impactful to the community and ecosystem. :)

The question is not ever "why not", it's "what would we have to not do in order to prioritize this feature.

**Comment by Prinsn** [comment]
The "in the first place" is answered by https://stackoverflow.com/a/1217757/2489966, to the degree that you can follow the specifics of the blog posts.

**Comment by RoyAwesome** [comment]
Hi, Sorry for the bump but I figured bumping this issue might be better than making a new issue one on this topic.

I have a slightly weird motivating example for wanting a yield return in a lambda, and frankly it's kind of an abuse of the IEnumerator system anyway.

Basically, I'm using the generator pattern to create a "Sequence" of events that happen.  For example:

```cs
   var MyActor = Whatever();
   var MyOtherActor = Whatever();
   var Seq = new Sequence(() => {
    DoSomething();
    yield return Wait.ForSeconds(.5);
    MyActor.MoveToLocation(MyOtherActor.Location);
    yield return Wait.ForTrue(() => MyActor.Location == MyOtherActor.Location);
   //and so on
  });

  AddSequenceToTicker(Seq)
```

In this example I create a Sequence object and hand it off to a ticker to run in an event loop.  Notably, I want to capture variables in the lambda so that I can do things with them while this sequence is going.  

Not being able to have a lambda here makes the resulting code extremely ugly, and it was a pretty big surprise to me to learn that I just wasn't able to do this as I expected.

I hope that some renewed interest may cause you all to re-evaluate making yield return work in lambdas.

**Reply by svick** [reply]
Wouldn't this work better using `await`, which is supported in lambdas?

```c#
   var Seq = new Sequence(async () => {
    DoSomething();
    await Wait.ForSeconds(.5);
    MyActor.MoveToLocation(MyOtherActor.Location);
    await Wait.ForTrue(() => MyActor.Location == MyOtherActor.Location);
   //and so on
  });
```

Note that `await` is not limited to `Task`s, so you can customize what the `await` actually does.

**Comment by bernd5** [comment]
As a workaround you can define a local function inside of your lambda, e.g.:

```csharp
using System;
using System.Collections.Generic;

Func<IEnumerable<int>> func = ()=>{
    return Do();
    IEnumerable<int> Do(){
        yield return 42;    
    }
};
```

**Reply by RoyAwesome** [reply]
that's... pretty ugly.  while it, and hoisting the named function out and using a named function at function level both work, they require whoever is creating a function like this to know that's the workaround for what should just work as all other specialty things in lambdas do (ie: async/await works in lambdas, why doesn't yield return?  they both suspend the function).

**Reply by CyrusNajmabadi** [reply]
> and frankly it's kind of an abuse of the IEnumerator system anyway.

It's not clear why it needs to not be "pretty ugly" if it's an admitted abuse anyways :-)

Why not just use async/await here?  You can customize it to be a super set of what yield does.

**Reply by RoyAwesome** [reply]
to be totally honest I didn't know you could.  That doesn't fix the surprising nature of learning you can't yield return in a lambda (but can in a named function).
